paasta_tools/api/tweens/request_logger.py

```diff
@@ -0,0 +1,117 @@
+# limitations under the License.
+"""
+Creates a tween that logs information about requests.
+"""
+import json
+import traceback
+from datetime import datetime
+
+import pyramid
+
+from paasta_tools.api import settings
+
+try:
+    import clog
+except ImportError:
+    clog = None
+
+
+DEFAULT_REQUEST_LOG_NAME = "tmp_paasta_api_requests"
+
+
+def includeme(config):
+    if clog is not None:
+        config.add_tween(
+            "paasta_tools.api.tweens.request_logger.request_logger_tween_factory",
+            under=pyramid.tweens.INGRESS,
+        )
+
+
+class request_logger_tween_factory:
+    """Tween that logs information about requests"""
+
+    def __init__(self, handler, registry):
+        self.handler = handler
+        self.registry = registry
+        self.log_name = registry.settings.get(
+            "request_log_name", DEFAULT_REQUEST_LOG_NAME,
+        )
+
+    def _log(
+        self, timestamp=None, level="INFO", additional_fields=None,
+    ):
+        if clog is not None:
+            # `settings` values are set by paasta_tools.api.api:setup_paasta_api
+            if timestamp is None:
+                timestamp = datetime.now()
+            dct = {
+                "human_timestamp": timestamp.strftime("%Y-%m-%dT%H:%M:%S"),
+                "unix_timestamp": timestamp.timestamp(),
+                "hostname": settings.hostname,
+                "level": level,
+                "cluster": settings.cluster,
+            }
+            if additional_fields is not None:
+                dct.update(additional_fields)
+            line = json.dumps(dct, sort_keys=True)
+            clog.log_line(self.log_name, line)
+
+    def __call__(self, request):
+        start_time = datetime.now()  # start clock for response time
+        request_fields = {
+            "path": request.path,
+            "params": request.params.mixed(),
+            "client_addr": request.client_addr,
+            "http_method": request.method,
+            "headers": dict(request.headers),  # incls user agent
+        }
+        response_fields = {}
+        log_level = "INFO"
+
+        try:
+            response = self.handler(request)
+
+            response_fields["status_code"] = response.status_int
+            if 300 <= response.status_int < 400:
+                log_level = "WARNING"
+            elif 400 <= response.status_int < 600:
+                log_level = "ERROR"
+                response_fields["body"] = response.body.decode("utf-8")
+
+            return response
+
+        except Exception as e:
+            log_level = "ERROR"
+            response_fields.update(
+                {
+                    "status_code": 500,
+                    "exc_type": type(e).__name__,
+                    "exc_info": traceback.format_exc(),
+                }
+            )
+            raise
+
+        finally:
+            response_time_ms = (datetime.now() - start_time).total_seconds() * 1000
```

Comment:
Why use datetime and then convert to seconds later? Doesn't time.time() work?
Also, this is minor, but why not just keep it as seconds instead of multiplying by 1000? We can always do unit conversions while analyzing the data.

Reply:
I switched to datetime to make it easier to add a human timestamp to the log lines. As for why not keep it as seconds, I kept it in line with most response time conventions and recorded it in milliseconds.